lib/index.ts

```diff
@@ -0,0 +1,572 @@
+			listConversations: _.partial(this.httpCall,
+				{ reqType: 'GET', path: 'inboxes/{inbox_id}/conversations[q{q}...][page={page}][limit={limit}]' }),
+			listTeammates: _.partial(this.httpCall,
+				{ reqType: 'GET', path: 'inboxes/{inbox_id}/teammates' }),
+		};
+		this.message = {
+			get: _.partial(this.httpCall,
+				{ reqType: 'GET', path: 'messages/{message_id}' }),
+			receiveCustom: _.partial(this.httpCall,
+				{ reqType: 'POST', path: 'channels/{channel_id}/incoming_messages' }),
+			reply: _.partial(this.httpCall,
+				{ reqType: 'POST', path: 'conversations/{conversation_id}/messages' }),
+			send: _.partial(this.httpCall,
+				{ reqType: 'POST', path: 'channels/{channel_id}/messages' }),
+		};
+	}
+
+	private httpCall = (details: FrontRequest, params: any, callback?: InternalCallback): Promise<any | void> => {
+		const requestOpts = {
+			body: (params) ? params : {},
+			headers: {
+				Authorization: `Bearer ${this.apiKey}`
+			},
+			json: true,
+			method: details.reqType,
+			url: `${FrontURL}/${this.formatPath(details.path, params)}`
+		};
+
+		// Make the request.
+		return request(requestOpts).promise().catch((error: any) => {
+			// Format this into something useful.
+			throw error.error._error;
+		}).asCallback(callback);
+	}
+
+	private formatPath(path: string, data: FrontRequestData | null): string {
+		const replaceTag = (substring: string) => {
+			// Test to see if this is an array type.
+			// If so, we need to repeat all entries in the array.
+			const preamble = substring.substring(1, substring.indexOf('{'));
+			let newString = substring;
+			let arraySubstitution = false;
+			if (substring.indexOf('...') !== -1) {
+				newString = _.replace(newString, /\.\.\./g, '');
+				arraySubstitution = true;
+			}
+			return _.replace(newString, /\{.*\}/g, (candidate: string) => {
+				// Search for the replacee in the data.
+				const replaceTrimmed = _.trimEnd(_.trim(candidate, '{'), '}');
+				if (testData[replaceTrimmed]) {
+					if (!arraySubstitution) {
+						return testData[replaceTrimmed];
+					} else {
+						let arrayString = '';
+						if (testData[replaceTrimmed].length === 0) {
+							return;
+						}
+
+						for (let index = 0; index < testData[replaceTrimmed].length; index += 1) {
+							let stringData = `${testData[replaceTrimmed][index]}`;
+							if (index > 0) {
+								stringData = `&${preamble}${data}`;
+							}
+							arrayString = arrayString.concat(stringData);
+						}
+
+						return arrayString;
+					}
+				}
+
+				return;
+			});
+		};
+		let optReplacements = 0;
+		let testData: any = (!data) ? {} : data;
+
+		// First replace any optional variables. If we don't find one, we
+		// remove this from the path.
+		// If we do find one, we replace with the appropriate parameterisation
+		// based on number already replaced.
+		const newPath = replaceTag(_.replace(path, /\[.*?\]/g, (candidate: string) => {
+			const queryChar = (optReplacements === 0) ? '?' : '&';
```

Comment:
Hmm. It possibly could, but I'd end up having to convert the string I'm chopping into a returned object to allow it to put it together again, and there are cases such as the 'q[someKey]=[someValue]&q[someOtherKey]=[someOtherValue]' where this would involve more work. I'm happy to do this in a next version when I don't feel so pressed, if it makes sense (and I look into how I'd do it a bit more).